package/libskylark/src/libskylark.c

```diff
@@ -0,0 +1,333 @@
+  return NO_ERROR;
+}
+
+RC get_accept_header(skylark_client_config_t* config)
+{
+  strcpy(config->accept_type_header, SBP_V2_ACCEPT_TYPE);
+  return NO_ERROR;
+}
+
+RC get_content_header(skylark_client_config_t* config)
+{
+  strcpy(config->content_type_header, SBP_V2_CONTENT_TYPE);
+  return NO_ERROR;
+}
+
+RC format_device_header(skylark_client_config_t* config)
+{
+  sprintf(config->device_header, DEVICE_UID_HEADER_FMT, config->device_uuid);
+  return NO_ERROR;
+}
+
+RC get_sbp_sender_id(u16* sender_id)
+{
+  *sender_id = SBP_SENDER_ID;
+  char sbp_sender_id_string[32];
+  if (file_read_string("/cfg/sbp_sender_id", sbp_sender_id_string,
+                       sizeof(sbp_sender_id_string)) < 0) {
+    return E_GENERIC_ERROR;
+
+  }
+  *sender_id = strtoul(sbp_sender_id_string, NULL, 10);
+  return NO_ERROR;
+}
+
+RC get_config(skylark_client_config_t* config)
+{
+  RC rc = NO_ERROR;
+  if ((rc = get_broker_endpoint(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = get_accept_header(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = get_content_header(config)) < NO_ERROR) {
+    return rc;
+  }
+  strcpy(config->encoding, STREAM_ENCODING);
+  if ((rc = get_device_uuid(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = format_device_header(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = get_sbp_sender_id(&config->sbp_sender_id)) < NO_ERROR) {
+    return rc;
+  }
+  return rc;
+}
+
+/**
+ *  Download (i.e., rover) processes.
+ */
+
+// Callback used by libcurl to pass data read from Skylark. Writes to pipe.
+//
+// TODO (mookerji): replace
+static size_t download_callback(void *p, size_t size, size_t n, void *up)
+{
+  int *fd = (int *)up;
+  ssize_t m = write(*fd, p, size * n);
+  return m;
+}
+
+RC download_process(skylark_client_config_t* config) {
+  CURLcode res = curl_global_init(CURL_GLOBAL_ALL);
+  if (res != CURLE_OK) {
+    exit(EXIT_FAILURE);
+  }
+  CURL *curl = curl_easy_init();
+  if (curl == NULL) {
+    curl_global_cleanup();
+    exit(EXIT_FAILURE);
+  }
+  curl_easy_setopt(curl, CURLOPT_URL, config->endpoint_url);
+  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, &download_callback);
+  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &config->fd_out);
+#ifdef VERBOSE
+  curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+#endif
+  curl_easy_setopt(curl, CURLOPT_USERAGENT, USER_AGENT);
+  struct curl_slist *chunk = NULL;
+  chunk = curl_slist_append(chunk, config->encoding);
+  chunk = curl_slist_append(chunk, config->accept_type_header);
+  chunk = curl_slist_append(chunk, config->device_header);
+  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, chunk);
+  res = curl_easy_perform(curl);
+  if (res != CURLE_OK) {
+    log_error(curl_easy_strerror(res));
+    curl_easy_cleanup(curl);
+    curl_global_cleanup();
+    exit(EXIT_FAILURE);
+  }
+  curl_easy_cleanup(curl);
+  curl_global_cleanup();
+  return NO_ERROR;
+}
+
+static void base_obs_callback(u16 sender_id, u8 len, u8 msg[], void *context)
```

Comment:
instead of having multiple callbacks here, should be able to make a generic callback (via a struct passed in context) to support multiple message types.